### PR TITLE
[Datasets] Autodetect dataset parallelism based on available resources and data size

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This PR defaults the parallelism of Dataset reads to `-1`. The parallelism is determined according to the following rule in this case:
- The number of available CPUs is estimated. If in a placement group, the number of CPUs in the cluster is scaled by the size of the placement group compared to the cluster size. If not in a placement group, this is the number of CPUs in the cluster. If the estimated CPUs is less than 8, it is set to 8.
- The parallelism is set to the estimated number of CPUs multiplied by 2.
- The in-memory data size is estimated. If the parallelism would create in-memory blocks larger than the target block size (512MiB), the parallelism is increased until the blocks are < 512MiB in size.

These rules fix two common user problems:
1. Insufficient parallelism in a large cluster, or too much parallelism on a small cluster.
2. Overly large block sizes leading to OOMs when processing a single block.

TODO:
- [x] Unit tests
- [x] Docs update

Supercedes part of: https://github.com/ray-project/ray/pull/25708

### CODE REVIEW

**Comment by ericl** [comment]
This is ready for review.

File: python/ray/data/read_api.py [hunk 1]
```diff
@@ -1053,11 +1063,73 @@ def _get_metadata(table: Union["pyarrow.Table", "pandas.DataFrame"]) -> BlockMet
...
+
+
+def _autodetect_parallelism(
+    parallelism: int, cur_pg: Optional[PlacementGroup], reader=None
+) -> int:
+    # Autodetect parallelism requested. The heuristic here are that we should try
+    # to create as many blocks needed to saturate available resources, and also keep
+    # block sizes below the target memory size, but no more. Creating too many
+    # blocks is inefficient.
+    if parallelism < 0:
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/read_api.py:1086]: The auto-detected parallelism will be truncated to the number of files in the datasource `prepare_read()` functions, which I think should be sufficient. Although we should note that this may result the target block size to be exceeded, but that will eventually be solved by dynamic block splitting.

File: python/ray/data/datasource/datasource.py [hunk 2]
```diff
@@ -100,6 +99,50 @@ def on_write_failed(
         pass
 
 
+@DeveloperAPI
+class Reader(Generic[T]):
```
Comment (clarkzinzow) [inline python/ray/data/datasource/datasource.py:103]: I'm concerned that introducing this `Reader` concept in addition to `Datasource` will make implementing custom datasources more difficult/confusing. The common case of subclassing `FileBasedDatasource` is covered since such subclasses can be `Reader`-agnostic, which is good, so this slight custom datasource UX degradation doesn't seem like a blocker.

It seems like the core reason that we can't do this read-time binding on the `Datasource` itself via its constructor (consolidating `Datasource` and `Reader`) is that the binding will need to be able to happen in a remote task, although we could work around this if we had `read_datasource` take an uninstantiated `Datasource`.

What do you think of this alternative, where the `Datasource` constructor is treated as the read-time binding? Note that this would be a breaking change for existing custom datasources.

```python
# datasource.py

class Datasource(Generic[T]):
    def estimate_inmemory_data_size(self) -> Optional[int]:
        return None

    def prepare_read(self, parallelism: int) -> List["ReadTask[T]"]:
        raise NotImplementedError

# file_based_datasource.py

class FileBasedDatasource(Datasource):
    def __init__(
        self,
        paths: Union[str, List[str]],
        filesystem: Optional["pyarrow.fs.FileSystem"] = None,
        schema: Optional[Union[type, "pyarrow.lib.Schema"]] = None,
        open_stream_args: Optional[Dict[str, Any]] = None,
        meta_provider: BaseFileMetadataProvider = DefaultFileMetadataProvider(),
        partition_filter: PathPartitionFilter = None,
        # TODO(ekl) deprecate this once read fusion is available.
        _block_udf: Optional[Callable[[Block], Block]] = None,
        **reader_args,
    ):
        _check_pyarrow_version()
        self._schema = schema
        self._open_stream_args = open_stream_args
        self._meta_provider = meta_provider
        self._partition_filter = partition_filter
        self._block_udf = _block_udf
        self._reader_args = reader_args
        paths, self._filesystem = _resolve_paths_and_filesystem(paths, filesystem)
        self._paths, self._file_sizes = meta_provider.expand_paths(
            paths, self._filesystem
        )

    def estimate_inmemory_data_size(self) -> Optional[int]:
        total_size = 0
        for sz in self._file_sizes:
            if sz is not None:
                total_size += sz
        return total_size

    def prepare_read(self, parallelism: int) -> List[ReadTask]:
        # ...

# read_api.py

@PublicAPI
def read_csv(
    paths: Union[str, List[str]],
    *,
    filesystem: Optional["pyarrow.fs.FileSystem"] = None,
    parallelism: int = 200,
    ray_remote_args: Dict[str, Any] = None,
    arrow_open_stream_args: Optional[Dict[str, Any]] = None,
    meta_provider: BaseFileMetadataProvider = DefaultFileMetadataProvider(),
    partition_filter: PathPartitionFilter = None,
    **arrow_csv_args,
) -> Dataset[ArrowRow]:
    return read_datasource(
        CSVDatasource,
        parallelism=parallelism,
        paths=paths,
        filesystem=filesystem,
        ray_remote_args=ray_remote_args,
        open_stream_args=arrow_open_stream_args,
        meta_provider=meta_provider,
        partition_filter=partition_filter,
        **arrow_csv_args,
    )

def read_datasource(
    datasource: Type[Datasource[T]],
    *,
    parallelism: int = -1,
    ray_remote_args: Dict[str, Any] = None,
    **read_args,
) -> Dataset[T]:
    # ...

    if force_local:
        read_tasks = _prepare_read(datasource, ctx, cur_pg, parallelism, read_args)
    else:
        # Prepare read in a remote task so that in Ray client mode, we aren't
        # attempting metadata resolution from the client machine.
        prepare_read = cached_remote_fn(
            _prepare_read, retry_exceptions=False, num_cpus=0
        )
        read_tasks = ray.get(
            prepare_read.remote(
                datasource,
                ctx,
                cur_pg,
                parallelism,
                read_args,
            )
        )

    # ...

def _prepare_read(
    ds_cls: Type[Datasource],
    ctx: DatasetContext,
    cur_pg: Optional[PlacementGroup],
    parallelism: int,
    kwargs: Dict[str, Any],
) -> List[ReadTask]:
    kwargs = _unwrap_arrow_serialization_workaround(kwargs)
    DatasetContext._set_current(ctx)
    ds = ds_cls(**kwargs)
    parallelism = _autodetect_parallelism(parallelism, cur_pg, ds)
    return ds.prepare_read(parallelism)
```

File: python/ray/data/datasource/datasource.py [hunk 3]
```diff
@@ -100,6 +99,50 @@ def on_write_failed(
         pass
 
 
+@DeveloperAPI
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/datasource/datasource.py:102]: Ah I thought it was public, yeah I think that we should make it public since we're advertising custom datasources as a feature.

File: python/ray/data/datasource/datasource.py (hunk 2 above)
Comment (ericl) [reply python/ray/data/datasource/datasource.py:103]: I thought about that, but I think the current approach is a little better in that:
1. We retain backwards compatibility, and
2. You get the flexibility of a single datasource class that can be used for multiple operations.

Probably either could work, but erring on the side of not breaking APIs.

Comment (clarkzinzow) [reply python/ray/data/datasource/datasource.py:103]: >You get the flexibility of a single datasource class that can be used for multiple operations.

Hmm but all existing datasources are constructed and used on a per-operation basis, and I can't immediately think of why we would want to reuse a datasource instance across multiple operations. That only makes sense if there's job/session-level data that the user wants to only supply once, and then apply to many read operations... It seems like this requirement might be a case of YAGNI.

> We retain backwards compatibility

I think that ensuring backwards compatibility is probably a good enough reason to keep the current `Reader` implementation. We'll just need to make sure that the custom datasource docs thoroughly explain `Datasource` vs `Reader`, and how to implement both.

File: python/ray/data/read_api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jianoaix) [reply python/ray/data/read_api.py:1107]: I think it's better to centralize these in one oracle. It's difficult to find where the parallelism got adjusted.

[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/read_api.py:1107]: The number of files isn't known here, so that is not a change we can do easily. It's in the datasource contract that the parallelism is just a "hint" and the actual datasource can decide a lower amount of parallelism if not feasible.

File: python/ray/data/datasource/datasource.py (hunk 2 above)
Comment (jianoaix) [reply python/ray/data/datasource/datasource.py:106]: Since we are prior GA, maybe breaking APIs is less a concern (than e.g. changing ray.remote) and will be worth it to get it right?

Comment (ericl) [reply python/ray/data/datasource/datasource.py:106]: Personally this pattern seems very ordinary / normal to me, so I'd push to keep it and also retain backwards compatibility. We should not be in the habit of breaking APIs unless there are very good reasons.

File: python/ray/data/read_api.py [hunk 4]
```diff
@@ -249,6 +259,7 @@ def read_datasource(
             prepare_read.remote(
                 datasource,
                 ctx,
+                cur_pg,
                 parallelism,
                 _wrap_and_register_arrow_serialization_workaround(read_args),
             )
```
Comment (jianoaix) [inline python/ray/data/read_api.py:265]: The check below needs to update (with the change to 1 CPU per task now etc.).
Also the warning log seems a bit off: it not necessarily has len(read_tasks) blocks since a read task can produce more than one blocks.

File: python/ray/data/datasource/datasource.py (hunk 2 above)
Comment (jianoaix) [reply python/ray/data/datasource/datasource.py:106]: Sounds good, after re-read it seems not a concern to me.

File: python/ray/data/read_api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jianoaix) [reply python/ray/data/read_api.py:1107]: Can the parallelism be tuned inside Reader? Like outside Reader, it's still the requested parallelism from user, and then there is a single place to tune/finalize the number inside Reader when all information is ready. It seems no use of parallelism outside Reader?

File: python/ray/data/datasource/datasource.py [hunk 5]
```diff
@@ -100,6 +102,50 @@ def on_write_failed(
...
+    """
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        """Return an estimate of the in-memory data size, or None if unknown.
+
+        Note that the in-memory data size may be larger than the on-disk data size.
+        """
+        raise NotImplementedError
+
+    def prepare_read(self, parallelism: int) -> List["ReadTask[T]"]:
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/datasource/datasource.py:121]: Renamed to get_read_tasks()!

File: python/ray/data/read_api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/read_api.py:1107]: It's currently tuned based on the number of files. Other readers might be reading from a database / REST API and may have other constraints, such as number of connections, etc., that cannot be guessed from outside the reader.

File: python/ray/data/datasource/datasource.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/datasource/datasource.py]: Changed to public.

**Comment by ericl** [comment]
Kicking off nightly tests in case this breaks some large scale dataset tests.

**Comment by ericl** [comment]
The hang in nightly tests seems to be our favorite: the parquet fragment serialization bug. This is since we return ReadTasks to the caller, and the read tasks have parquet fragments in it.

@clarkzinzow , what do you think about registering the parquet fragment serializer on import and not de-registering it. The issue seems to be (1) we didn't register it in read_api.py, and (2) it's deregistered prior to ReadTasks getting returned from a remote function, hence bypassing the serializer. An alternative is for the parquet ReadTasks to explicitly call the serializer.

**Comment by scv119** [comment]
@jianoaix @clarkzinzow can we prioritize this?

File: python/ray/data/context.py [hunk 6]
```diff
@@ -11,7 +11,7 @@
 _context_lock = threading.Lock()
 
 # The max target block size in bytes for reads and transformations.
-DEFAULT_TARGET_MAX_BLOCK_SIZE = 2048 * 1024 * 1024
+DEFAULT_TARGET_MAX_BLOCK_SIZE = 512 * 1024 * 1024
```
Comment (stephanie-wang) [inline python/ray/data/context.py:14]: Is it possible to make this user-configurable per dataset?

Comment (ericl) [reply python/ray/data/context.py:14]: Yes, you can configure it via the DatasetContext.

File: python/ray/data/read_api.py [hunk 7]
```diff
@@ -1072,12 +1090,86 @@ def _get_metadata(table: Union["pyarrow.Table", "pandas.DataFrame"]) -> BlockMet
...
+        if parallelism != -1:
+            raise ValueError("`parallelism` must either be -1 or a positive integer.")
+        # Start with 2x the number of cores as a baseline, with a min floor.
+        avail_cpus = _estimate_avail_cpus(cur_pg)
+        parallelism = max(ctx.min_parallelism, avail_cpus * 2)
+        if reader:
+            # Increase it to avoid overly-large blocks as needed.
+            mem_size = reader.estimate_inmemory_data_size()
+            if mem_size is not None:
+                parallelism = max(
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/read_api.py:1137]: We already do that later (line 278).

File: python/ray/data/context.py (hunk 6 above)
Comment (stephanie-wang) [reply python/ray/data/context.py:14]: Ah I see. It would be nice if the user could pass it as a top-level arg like `parallelism`, but maybe we can do this as a follow-up?

File: python/ray/data/read_api.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/read_api.py:1137]: Good point, let me add this.

File: python/ray/data/context.py (hunk 6 above)
Comment (ericl) [reply python/ray/data/context.py:14]: I see, agree that could be nice. I think we should do this when enabling block splitting by default (and maybe deprecating parallelism at the same time. That way, the story becomes pretty simple: parallelism is auto-detected usually, and the block size can be reduced if needed if you're running into OOMs.

**Comment by ericl** [comment]
@stephanie-wang updated to add the suggested warning as well.

**Comment by ericl** [comment]
Btw, the nightly tests are still failing on pipelined ingest.

File: python/ray/data/read_api.py [hunk 8]
```diff
@@ -238,23 +250,41 @@ def read_datasource(
...
+                cur_pg,
                 parallelism,
                 _wrap_and_register_arrow_serialization_workaround(read_args),
             )
         )
 
-    if len(read_tasks) < parallelism and (
+    if read_tasks and len(read_tasks) < min_safe_parallelism:
+        perc = 1 + round((min_safe_parallelism - len(read_tasks)) / len(read_tasks), 1)
+        logger.warning(
```
Comment (stephanie-wang) [inline python/ray/data/read_api.py:279]: Nice warning message! Nit, we may want to add some margin to the min_safe_parallelism so we don't print 1.0x when the block size is really close.

**Review by stephanie-wang — APPROVED** [review]
It would be great if we could also pytest the logger warnings.

**Review by jianoaix — APPROVED** [review]
The nightly passed!

**Comment by clarkzinzow** [comment]
Ah so it looked like the real issue was that `self` was captured in the `read_pieces` closure, which contained a reference to the `ParquetDataset`, so every read task was deserializing every piece/fragment in the Arrow `ParquetDataset?

File: doc/source/data/performance-tips.rst [hunk 9]
```diff
@@ -109,13 +109,13 @@ This can be used in conjunction with column pruning when appropriate to get the
 Tuning Read Parallelism
 ~~~~~~~~~~~~~~~~~~~~~~~
 
-By default, Ray requests 0.5 CPUs per read task, which means two read tasks can concurrently execute per CPU.
+By default, Ray requests 1 CPU per read task, which means one read tasks can concurrently execute per CPU.
```
Comment (clarkzinzow) [inline doc/source/data/performance-tips.rst:112]:
```suggestion
By default, Ray requests 1 CPU per read task, which means one read task per CPU can execute concurrently.
```

File: python/ray/data/_internal/progress_bar.py [hunk 10]
```diff
@@ -87,7 +87,7 @@ def set_description(self, name: str) -> None:
             self._bar.set_description(name)
 
     def update(self, i: int) -> None:
-        if self._bar:
+        if self._bar and i != 0:
```
Comment (clarkzinzow) [inline python/ray/data/_internal/progress_bar.py:90]: Why is this change needed?

File: python/ray/data/datasource/file_based_datasource.py [hunk 11]
```diff
@@ -429,6 +318,138 @@ def file_extension_filter(cls) -> Optional[PathPartitionFilter]:
...
+                    # Pass Snappy compression as a reader arg, so datasource subclasses
+                    # can manually handle streaming decompression in
+                    # self._delegate._read_stream().
+                    reader_args["compression"] = compression
+                    reader_args["filesystem"] = fs
+                elif compression is not None:
+                    # Non-Snappy compression, pass as open_input_stream() arg so Arrow
+                    # can take care of streaming decompression for us.
+                    open_stream_args["compression"] = compression
+                with self._delegate._open_input_source(
```
Comment (clarkzinzow) [inline python/ray/data/datasource/file_based_datasource.py:418]: We shouldn't capture `self` in this closure, since that's going to create serialization issues, e.g. the `self.filesystem` deserialization might break since it's not using our wrapper. We should change `FileBasedDatasource._open_input_source()` to a static method (it's not currently using the instance anyway) and close over that function pointer instead of `self`.

File: python/ray/data/datasource/parquet_datasource.py [hunk 12]
```diff
@@ -36,38 +37,30 @@
 FILE_READING_RETRY = 8
 
 
-def _register_parquet_file_fragment_serialization():
-    from pyarrow.dataset import ParquetFileFragment
+# TODO(ekl) this is a workaround for a pyarrow serialization bug, where serializing a
+# raw pyarrow file fragment causes S3 network calls.
```
Comment (clarkzinzow) [inline python/ray/data/datasource/parquet_datasource.py:41]: Won't this no longer be an issue since we're no longer closing over `self`, which contained the full `ParquetDataset` object? It seems like that was the real source of the issue with the failing nightly test.

File: python/ray/data/read_api.py [hunk 13]
```diff
@@ -238,23 +250,37 @@ def read_datasource(
...
                 datasource,
                 ctx,
+                cur_pg,
                 parallelism,
                 _wrap_and_register_arrow_serialization_workaround(read_args),
             )
         )
 
-    if len(read_tasks) < parallelism and (
+    if read_tasks and len(read_tasks) < min_safe_parallelism * 0.7:
```
Comment (clarkzinzow) [inline python/ray/data/read_api.py:273]: Why this 0.7 factor?

**Comment by jianoaix** [comment]
Hmm I might be in a different version....let me clone this and test again.

**Review by clarkzinzow — DISMISSED** [review]
LGTM overall, but we should not be capturing `self` in the `read_files` function in `_FileBasedDatasourceReader.get_read_tasks()`, due to potential serialization issues.

File: python/ray/data/read_api.py (hunk 13 above)
Comment (ericl) [reply python/ray/data/read_api.py:273]: This was to avoid warnings in reasonable OK cases.

File: python/ray/data/datasource/parquet_datasource.py (hunk 12 above)
Comment (ericl) [reply python/ray/data/datasource/parquet_datasource.py:41]: You still get RPCs serializing the fragments, which are also passed separately independently of self.

File: python/ray/data/_internal/progress_bar.py (hunk 10 above)
Comment (ericl) [reply python/ray/data/_internal/progress_bar.py:90]: This avoids progress bar update spam where the bar isn't actually moving (it cleans up non interactive shells).

File: python/ray/data/datasource/parquet_datasource.py (hunk 12 above)
Comment (clarkzinzow) [reply python/ray/data/datasource/parquet_datasource.py:41]: I'm not saying that we can get rid of all serialization hacks for file fragments, I'm asking if the old method (registering a serialization handler) actually was working as intended for the file fragments that are passed as an argument and that closing over `self` was the issue, which would bypass either of these serialization hacks.

In other words, I'm wondering if we revert back to the old serialization hack of registering a serialization handler (i.e. reverting this commit), whether the nightly test would pass.

Comment (ericl) [reply python/ray/data/datasource/parquet_datasource.py:41]: Possibly, but I'd rather handle things explicitly the same way everywhere rather than debug registration issues. It's unclear what issues would remain with reverting that commit.

**Comment by jianoaix** [comment]
Spent a bit time wrestling the cluster setup for testing, but got the result now, which passed the test: https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_SXULguMtWTFYrRLXwDb1kaYB

I'll work on fixing the CI / addressing the comments (since Eric is OOO).

File: python/ray/data/datasource/parquet_datasource.py (hunk 12 above)
Comment (jianoaix) [reply python/ray/data/datasource/parquet_datasource.py:41]: IMO this class is easier to understand than using registration. Registering and then unregistering right after the code block seems not intuitive to me (the registry may be expected to last the lifecycle of process).

The most confusing part is indeed the RPCs inside these ser/deser. Just to be clear about this: does this make RPCs both in serialization and deserialization? (deserialization probably does because we do retry to fix HDFS issue before).

Comment (clarkzinzow) [reply python/ray/data/datasource/parquet_datasource.py:41]: >Registering and then unregistering right after the code block seems not intuitive to me (the registry may be expected to last the lifecycle of process).

Yeah we shouldn't be doing that for the registration pattern, that's not what we do for the JSON read options, for example.

>The most confusing part is indeed the RPCs inside these ser/deser. Just to be clear about this: does this make RPCs both in serialization and deserialization? (deserialization probably does because we do retry to fix HDFS issue before).

The network requests only happen on deserialization AFAIK.

File: python/ray/data/datasource/file_based_datasource.py [hunk 14]
```diff
@@ -429,6 +318,138 @@ def file_extension_filter(cls) -> Optional[PathPartitionFilter]:
...
+                    # Pass Snappy compression as a reader arg, so datasource subclasses
+                    # can manually handle streaming decompression in
+                    # self._delegate._read_stream().
+                    reader_args["compression"] = compression
+                    reader_args["filesystem"] = fs
+                elif compression is not None:
+                    # Non-Snappy compression, pass as open_input_stream() arg so Arrow
+                    # can take care of streaming decompression for us.
+                    open_stream_args["compression"] = compression
+                with FileBasedDatasource._open_input_source(
```
[the remark this answers is not in the document]
Comment (jianoaix) [reply python/ray/data/datasource/file_based_datasource.py]: Good catch!

**Comment by ericl** [comment]
test_tensors_shuffle failing

File: python/ray/data/datasource/parquet_datasource.py [hunk 15]
```diff
@@ -243,69 +180,138 @@ def read_pieces(serialized_pieces: str) -> Iterator[pa.Table]:
...
+        self._metadata = meta_provider.prefetch_file_metadata(pq_ds.pieces) or []
+        self._pq_ds = pq_ds
+        self._meta_provider = meta_provider
+        self._inferred_schema = inferred_schema
+        self._block_udf = _block_udf
+        self._reader_args = reader_args
+        self._columns = columns
+        self._schema = schema
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
```
Comment (c21) [inline python/ray/data/datasource/parquet_datasource.py:192]: sorry for the late comment. but I think it's probably a bug here to rely on `serialized_size`, which seems to be size of file footer, not size of actual data. Crafted a quick fix in https://github.com/ray-project/ray/pull/26516 , please let me know if it makes sense or not, thanks.
